AntennaTracker/control_auto.cpp

```diff
@@ -58,3 +70,57 @@ void Tracker::calc_body_frame_target(float pitch, float yaw, float& bf_pitch, fl
     bf_pitch   = ahrs.cos_roll()     * pitch    + ahrs.sin_roll()  * ahrs.cos_pitch() * yaw;
     bf_yaw     = -ahrs.sin_roll()    * pitch    + ahrs.cos_pitch() * ahrs.cos_roll()  * yaw;
 }
+bool Tracker::convert_bf_to_ef(float pitch, float yaw, float& ef_pitch, float& ef_yaw)
+{
+    // avoid divide by zero
+    if (is_zero(ahrs.cos_pitch())) {
+        return false;
+    }
+    // convert earth frame angle or rates to body frame
+    ef_pitch =  ahrs.cos_roll()                     * pitch -  ahrs.sin_roll()                     * yaw; //pitch
+    ef_yaw =   (ahrs.sin_roll() / ahrs.cos_pitch()) * pitch + (ahrs.cos_roll() / ahrs.cos_pitch()) * yaw; //yaw
+    return true;
+}
+void Tracker::set_servo_direction(bool& direction_reversed)
+{
+    // calculating distances from current pitch/yaw to lower and upper limits in centi-degrees
+    float yaw_angle_limit_lower =   100 * (channel_yaw.get_servo_out()-(-9000))  / (9000-(-9000)) * g.yaw_range;
+    float yaw_angle_limit_upper =   100 * (9000-channel_yaw.get_servo_out())    / (9000-(-9000)) * g.yaw_range;
+    float pitch_angle_limit_lower = 100 * (channel_pitch.get_servo_out()-(-9000)) / (9000-(-9000))   * (g.pitch_max - g.pitch_min);
+    float pitch_angle_limit_upper = 100 * (9000-channel_pitch.get_servo_out())   / (9000-(-9000))   * (g.pitch_max - g.pitch_min);
+
+    // distances to earthframe angle limits in centi-degrees
+    float ef_yaw_limit_lower = yaw_angle_limit_lower;
+    float ef_yaw_limit_upper = yaw_angle_limit_upper;
+    float ef_pitch_limit_lower = pitch_angle_limit_lower;
+    float ef_pitch_limit_upper = pitch_angle_limit_upper;
+    if (convert_bf_to_ef(pitch_angle_limit_lower, yaw_angle_limit_lower, ef_pitch_limit_lower, ef_yaw_limit_lower)){}
+    if (convert_bf_to_ef(pitch_angle_limit_upper, yaw_angle_limit_upper, ef_pitch_limit_upper, ef_yaw_limit_upper)){}
+
+    float ef_yaw_current = wrap_180_cd(ahrs.yaw_sensor);
+    float ef_yaw_target = wrap_180_cd((nav_status.bearing+g.yaw_trim)*100);
+    float ef_yaw_angle_error = wrap_180_cd(ef_yaw_target - ef_yaw_current);
+
+    // sets lower and upper angle errors
+    float upper_angle_error;
```

Comment:
It might be good to rename "upper_angle_error" to "ef_yaw_angle_error_upper" to make it more clear that it's in the same frame and axis as "ef_yaw_angle_error".